py-tensorflow: submission 1.4.1

emcrisostomo opened the pull request:
#### Description

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->
<!-- (delete all below for minor changes) -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [ ] enhancement
- [ ] security fix
- [x] submission

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
macOS 10.13.2
Xcode 9.2

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

jmroot commented on python/py-tensorflow-tensorboard/Portfile:
```diff
@@ -0,0 +1,68 @@
+                    and understanding your TensorFlow runs and graphs.
+
+homepage            https://github.com/tensorflow/tensorboard
+master_sites        https://pypi.python.org/packages/d5/fb/80df4eb3234c41beeeb3122b1966effbf08608fb80d2fa74bb72f8da9cb3
```
Does `pypi:/t/tensorflow-tensorboard` not work here? The files.pythonhosted.org redirector should usually work to find the appropriate hex string subdir.

emcrisostomo replied:
Nope, I tried to make it work but with no luck. I think it's related to its being a `whl`. Look at the attempt:

```--->  Fetching distfiles for py27-tensorflow-tensorboard
--->  Attempting to fetch tensorflow_tensorboard-0.4.0rc3-py2-none-any.whl from https://files.pythonhosted.org/packages/source/t/tensorflow-tensorboard
--->  Attempting to fetch tensorflow_tensorboard-0.4.0rc3-py2-none-any.whl from https://pypi.python.org/packages/source/t/tensorflow-tensorboard
--->  Attempting to fetch tensorflow_tensorboard-0.4.0rc3-py2-none-any.whl from http://lil.fr.distfiles.macports.org/py-tensorflow-tensorboard
... snip ...
Error: Failed to fetch py27-tensorflow-tensorboard: ... snip ...```

jmroot commented on python/py-tensorflow-tensorboard/Portfile:
```diff
@@ -0,0 +1,68 @@
+extract.cmd         cp
+extract.pre_args
+extract.post_args   ${worksrcpath}
+extract.mkdir       yes
```
This could all probably be avoided if you told pip to install the wheel from `${distpath}/${distfiles}` in destroot.args below.

emcrisostomo replied:
Ah ok. Can I then skip the extract phase with a simple:

``` extract {
}```

?

jmroot replied:
Yes.

emcrisostomo replied:
Done. Thanks, it made the trick

jmroot commented on python/py-tensorflow-tensorboard/Portfile:
```diff
@@ -0,0 +1,68 @@
+
+    if {${python.version} eq 27} {
+        depends_lib-append \
+            port:py${python.version}-futures \
```
Trailing backslash needs to be removed.

emcrisostomo replied:
Right, just pushed.

jmroot commented on python/py-tensorflow/Portfile:
```diff
@@ -0,0 +1,82 @@
+homepage            https://www.tensorflow.org/
+
+if {${python.version} eq 27} {
+    master_sites        https://pypi.python.org/packages/eb/40/0c425a97870f0e95a9d79e4fd5c2f609ee89a7e69b4644579d3157ed5ea0
```
Same question as for py-tensorflow-tensorboard, does using `pypi:t/tensorflow` not work? That would also have the advantage of not having to be set differently for each subport.

emcrisostomo replied:
See above: I tried but no luck with the `whl` file

jmroot commented on python/py-tensorflow-tensorboard/Portfile:
```diff
@@ -0,0 +1,68 @@
+    destroot.pre_args
+    destroot.args       \
+        install --no-cache-dir --no-dependencies \
+        --target ${destroot}${python.pkgd} \
```
Is this correct? I would have thought you would install into a destroot with `--root=${destroot}`, which is what the python portgroup normally puts into ${destroot.destdir} (which is passed in destroot.post_args except you're clearing that below.)

emcrisostomo replied:
`--root` is the first thing I tried and I hit strange permission errors. I retried after your review and it seems to work. At this point I may have had some garbage, I can't replicate the error now.

emcrisostomo replied:
Ok, it wasn't apparent just changing one port. If I install the `py-tensorflow`port when using `--root=${destroot}` I get this behaviour:

```
% sudo port install py-tensorflow
--->  Computing dependencies for py-tensorflow
The following dependencies will be installed:
 py27-tensorflow
 py27-tensorflow-tensorboard
Continue? [Y/n]: y
--->  Fetching distfiles for py27-tensorflow-tensorboard
--->  Verifying checksums for py27-tensorflow-tensorboard
--->  Extracting py27-tensorflow-tensorboard
--->  Configuring py27-tensorflow-tensorboard
--->  Building py27-tensorflow-tensorboard
--->  Staging py27-tensorflow-tensorboard into destroot
--->  Installing py27-tensorflow-tensorboard @0.4.0rc3_0
--->  Activating py27-tensorflow-tensorboard @0.4.0rc3_0
--->  Cleaning py27-tensorflow-tensorboard
--->  Fetching distfiles for py27-tensorflow
--->  Verifying checksums for py27-tensorflow
--->  Extracting py27-tensorflow
--->  Configuring py27-tensorflow
--->  Building py27-tensorflow
--->  Staging py27-tensorflow into destroot
--->  Installing py27-tensorflow @1.4.1_0
--->  Activating py27-tensorflow @1.4.1_0
Error: Failed to activate py27-tensorflow: Image error: /opt/local/Library/Frameworks/Python.framework/Versions/2.7/bin/tensorboard is being used by the active py27-tensorflow-tensorboard port.  Please deactivate this port first, or use 'port -f activate py27-tensorflow' to force the activation.
Error: See /opt/local/var/macports/logs/_Users_enrico_ports_python_py-tensorflow/py27-tensorflow/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets to report a bug.
Error: Processing of port py-tensorflow failed

% sudo port -f activate py27-tensorflow
--->  Computing dependencies for py27-tensorflow
--->  Activating py27-tensorflow @1.4.1_0
Warning: File /opt/local/Library/Frameworks/Python.framework/Versions/2.7/bin/tensorboard already exists.  Moving to: /opt/local/Library/Frameworks/Python.framework/Versions/2.7/bin/tensorboard.mp_1513806190.
Warning: File /opt/local/bin/tensorboard-2.7 already exists.  Moving to: /opt/local/bin/tensorboard-2.7.mp_1513806190.
--->  Cleaning py27-tensorflow
```

If I use `--target`, then I get this behaviour:

```
% sudo port install py-tensorflow
--->  Computing dependencies for py-tensorflow
The following dependencies will be installed:
 py27-tensorflow
 py27-tensorflow-tensorboard
Continue? [Y/n]: y
--->  Fetching distfiles for py27-tensorflow-tensorboard
--->  Verifying checksums for py27-tensorflow-tensorboard
--->  Extracting py27-tensorflow-tensorboard
--->  Configuring py27-tensorflow-tensorboard
--->  Building py27-tensorflow-tensorboard
--->  Staging py27-tensorflow-tensorboard into destroot
--->  Installing py27-tensorflow-tensorboard @0.4.0rc3_0
--->  Activating py27-tensorflow-tensorboard @0.4.0rc3_0
--->  Cleaning py27-tensorflow-tensorboard
--->  Fetching distfiles for py27-tensorflow
--->  Verifying checksums for py27-tensorflow
--->  Extracting py27-tensorflow
--->  Configuring py27-tensorflow
--->  Building py27-tensorflow
--->  Staging py27-tensorflow into destroot
--->  Installing py27-tensorflow @1.4.1_0
--->  Activating py27-tensorflow @1.4.1_0
--->  Cleaning py27-tensorflow
--->  Fetching distfiles for py-tensorflow
--->  Verifying checksums for py-tensorflow
--->  Extracting py-tensorflow
--->  Configuring py-tensorflow
--->  Building py-tensorflow
--->  Staging py-tensorflow into destroot
--->  Installing py-tensorflow @1.4.1_0
--->  Activating py-tensorflow @1.4.1_0
--->  Cleaning py-tensorflow
--->  Updating database of binaries
--->  Scanning binaries for linking errors
--->  No broken files found.
```

I have no idea why, but I banged my head against the table quite a few times a couple of days ago.

jmroot commented on python/py-tensorflow-tensorboard/Portfile:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    destroot.cmd        pip-${python.branch}
+    destroot.pre_args
```
This would normally contain ${destroot.target}, which is `install` by default. So you could just leave that out of destroot.args instead.

emcrisostomo replied:
Thanks, I had no idea and it's cleaner.

emcrisostomo commented:
@jmroot, I pushed the changes you suggested after having tried them locally. I'm not sure how I can fix this error though:

```
Error: Failed to activate py27-tensorflow: Image error: /opt/local/Library/Frameworks/Python.framework/Versions/2.7/bin/tensorboard is being used by the active py27-tensorflow-tensorboard port.  Please deactivate this port first, or use 'port -f activate py27-tensorflow' to force the activation.
```

jmroot commented:
Apparently both packages install that file? You can't have two ports that install the same file active simultaneously. Need to make it so only one of the ports installs it. (You can simply delete files from the destroot in a post-destroot block if all else fails.)

emcrisostomo commented:
Ok, I will check whether those two files are identical. If they are (I remember it's just a thin shell script wrapper to forward arguments to python), I'll drop one (probably the one in `py-tensorflow`).

emcrisostomo commented:
@jmroot, the port looks fine to me now. I applied all your suggestions and squashed the commits back to one per port.

jmroot commented on python/py-backports.weakref/Portfile:
```diff
@@ -0,0 +1,30 @@
+name                py-backports.weakref
+version             1.0.post1
+platforms           darwin
+license             Python Software Foundation License
```
This should be `PSF`.

emcrisostomo replied:
@jmroot Done, thanks.

emcrisostomo commented:
Can this PR proceed or is there any more work to do?

pmetzger commented:
@jmroot ?

jmroot commented:
Probably need to look at whether the new ports are setting supported_archs correctly? And support archs other than x86_64 if possible (OK if it's not).

pmetzger commented:
@emcrisostomo So we're very close here. If you can just look into that last bit we can merge.

pmetzger commented:
@emcrisostomo Oh, and if you can squash when you do it...

pmetzger commented:
@emcrisostomo ?

emcrisostomo commented:
Sorry @pmetzger but I was fighting against a strange problem. Finally, I solved it: basically, this port depends on `python/py-protobuf3` and it has been working fine until `devel/protobuf3-cpp` has been updated. After that, suddenly, new installations of this port started segfaulting. I'm not sure whether `python/py-protobuf3` should be kept aligned with `devel/protobuf3-cpp`.

To summarise:

  * I updated `python/py-protobuf3` to 3.5.1.
  * I squashed the commits.
  * I rebased the pull request over the latest `master` since `devel/protobuf3-cpp` has already been updated to 3.5.1 in `master`.

Now it works fine: I tested an upgrade and two vanilla installations and it can finally be merged.

jmroot commented:
It looks like py-tensorflow-tensorboard and py-backports.weakref are noarch? If so they should say `supported_archs noarch`.

emcrisostomo commented:
@jmroot I've added `supported_archs` to the two ports you mention.

pmetzger commented:
@emcrisostomo Can you squash again?  @jmroot are you happy/can we commit?
